File: tv_alerts/analysis.py
"""Confirm whether a TradingView "Crossing" alert was a real resistance
breakout backed by rising volume, versus noise.

A single price-cross alert only tells you TradingView saw the price touch
a level intraday. It says nothing about:
  - whether the level was actually a meaningful resistance (a recent swing
    high) rather than an arbitrary number,
  - whether price *closed* above it (a wick through and back is not a
    breakout),
  - whether volume confirmed the move, and
  - whether the breakout held on subsequent days rather than failing.

`analyze_alert` checks all four using daily OHLCV bars around the alert.
"""

from __future__ import annotations

from dataclasses import dataclass

from tv_alerts.models import Alert, AlertType, BreakoutVerdict, PriceBar


@dataclass
class AnalysisParams:
    # How many prior trading days to average for the "normal" volume baseline.
    baseline_days: int = 20
    # Breakout-day volume must be at least this multiple of the baseline
    # average to count as confirmed.
    volume_ratio_threshold: float = 1.5
    # Window (in trading days, including the breakout day) used to check
    # that volume was trending up into the breakout, not just a lone spike.
    volume_trend_days: int = 3
    # How far back to look for the swing high that makes `level` a
    # meaningful resistance rather than an arbitrary crossing.
    recent_high_lookback_days: int = 60
    # `level` counts as "near/above the recent high" if it's within this
    # fraction below the highest high seen in the lookback window.
    recent_high_tolerance: float = 0.02
    # Number of subsequent trading days the close must stay above `level`
    # to call the breakout "held".
    hold_days: int = 2
    # Minimum bars of prior history required before we'll trust the
    # baseline/recent-high calculations at all.
    min_baseline_bars: int = 10


def _find_alert_bar_index(bars: list[PriceBar], alert: Alert) -> int | None:
    alert_date = alert.time.date()
    for i, bar in enumerate(bars):
        if bar.date.date() >= alert_date:
            return i
    return None


def analyze_alert(
    alert: Alert, bars: list[PriceBar], params: AnalysisParams | None = None
) -> BreakoutVerdict:
    params = params or AnalysisParams()
    bars = sorted(bars, key=lambda b: b.date)

    if alert.alert_type != AlertType.PRICE_CROSS or alert.level is None:
        return BreakoutVerdict(
            alert=alert,
            close_on_alert_day=None,
            pct_above_level=None,
            volume_on_alert_day=None,
            avg_volume_baseline=None,
            volume_ratio=None,
            volume_trend_ratio=None,
            near_recent_high=None,
            held_above_level=None,
            days_held=0,
            verdict="SKIPPED",
            notes=f"Not a numeric price-level alert ({alert.alert_type.value}).",
        )

    idx = _find_alert_bar_index(bars, alert)
    if idx is None:
        return _no_data_verdict(alert, "No bar found on or after the alert date.")

    breakout_bar = bars[idx]
    baseline = bars[max(0, idx - params.baseline_days) : idx]
    if len(baseline) < params.min_baseline_bars:
        return _no_data_verdict(
            alert,
            f"Only {len(baseline)} prior trading day(s) of history available "
            f"(need {params.min_baseline_bars}).",
        )

    avg_volume_baseline = sum(b.volume for b in baseline) / len(baseline)
    volume_ratio = (
        breakout_bar.volume / avg_volume_baseline if avg_volume_baseline > 0 else None
    )

    trend_days = params.volume_trend_days
    recent_window = bars[max(0, idx - trend_days + 1) : idx + 1]
    prior_window = bars[max(0, idx - 2 * trend_days + 1) : max(0, idx - trend_days + 1)]
    volume_trend_ratio = None
    if prior_window:
        prior_avg = sum(b.volume for b in prior_window) / len(prior_window)
        recent_avg = sum(b.volume for b in recent_window) / len(recent_window)
        if prior_avg > 0:
            volume_trend_ratio = recent_avg / prior_avg

    high_window = bars[max(0, idx - params.recent_high_lookback_days) : idx]
    recent_high = max((b.high for b in high_window), default=None)
    near_recent_high = (
        recent_high is not None
        and alert.level >= recent_high * (1 - params.recent_high_tolerance)
    )

    close_on_alert_day = breakout_bar.close
    pct_above_level = (close_on_alert_day - alert.level) / alert.level * 100

    following = bars[idx + 1 : idx + 1 + params.hold_days]
    days_held = 0
    for bar in following:
        if bar.close > alert.level:
            days_held += 1
        else:
            break
    held_above_level: bool | None
    if len(following) < params.hold_days:
        held_above_level = None  # not enough time has passed yet to know
    else:
        held_above_level = days_held == params.hold_days

    closed_above = close_on_alert_day > alert.level
    volume_confirmed = volume_ratio is not None and volume_ratio >= params.volume_ratio_threshold
    volume_growing = volume_trend_ratio is None or volume_trend_ratio >= 1.0

    notes_parts = [
        f"close {'above' if closed_above else 'at/below'} level "
        f"({pct_above_level:+.1f}%)",
        f"volume {volume_ratio:.2f}x {params.baseline_days}d avg" if volume_ratio else "volume ratio n/a",
    ]
    if volume_trend_ratio is not None:
        notes_parts.append(f"{trend_days}d volume trend {volume_trend_ratio:.2f}x")
    notes_parts.append(
        "near/above recent high" if near_recent_high else "well below recent high (weak resistance)"
    )
    if held_above_level is True:
        notes_parts.append(f"held above level for {days_held}/{params.hold_days}d")
    elif held_above_level is False:
        notes_parts.append(f"failed to hold - closed back below within {params.hold_days}d")
    else:
        notes_parts.append("not enough time elapsed yet to confirm it held")

    if not closed_above:
        verdict = "NO_CLOSE_CONFIRM"
    elif volume_confirmed and near_recent_high and volume_growing and held_above_level is not False:
        verdict = "CONFIRMED_BREAKOUT"
    elif volume_confirmed and near_recent_high:
        verdict = "WATCH"
    elif volume_confirmed or near_recent_high:
        verdict = "WATCH_WEAK"
    else:
        verdict = "NO"

    return BreakoutVerdict(
        alert=alert,
        close_on_alert_day=close_on_alert_day,
        pct_above_level=pct_above_level,
        volume_on_alert_day=breakout_bar.volume,
        avg_volume_baseline=avg_volume_baseline,
        volume_ratio=volume_ratio,
        volume_trend_ratio=volume_trend_ratio,
        near_recent_high=near_recent_high,
        held_above_level=held_above_level,
        days_held=days_held,
        verdict=verdict,
        notes="; ".join(notes_parts),
    )


def _no_data_verdict(alert: Alert, reason: str) -> BreakoutVerdict:
    return BreakoutVerdict(
        alert=alert,
        close_on_alert_day=None,
        pct_above_level=None,
        volume_on_alert_day=None,
        avg_volume_baseline=None,
        volume_ratio=None,
        volume_trend_ratio=None,
        near_recent_high=None,
        held_above_level=None,
        days_held=0,
        verdict="INSUFFICIENT_DATA",
        notes=reason,
    )

File: src/analysis.ts
/**
 * Confirm whether a price-level crossing was a real move through the level,
 * backed by volume, versus noise.
 *
 * A single price-cross alert only tells you price touched a level intraday.
 * It says nothing about:
 *   - whether the level sits at a recent extreme (a swing high for an upward
 *     crossing, a swing low for a downward one) rather than being an arbitrary
 *     number in the middle of the range,
 *   - whether price *closed* past it (a wick through and back is not a move),
 *   - whether volume confirmed the move, and
 *   - whether closes stayed past the level on subsequent days.
 *
 * `analyzeAlert` checks all four using daily OHLCV bars around the alert, in
 * the direction the alert crossed (`Alert.direction`, absent = up). The
 * verdict names and BreakoutVerdict field names are upside-flavoured for
 * historical reasons and kept because they are persisted; for a downward
 * crossing `nearRecentHigh` means near the recent low and `heldAboveLevel`
 * means held below.
 */

import type { Alert, BreakoutVerdict, PriceBar } from "./models.js";

export interface AnalysisParams {
  /** How many prior trading days to average for the "normal" volume baseline. */
  baselineDays: number;
  /** Breakout-day volume must be at least this multiple of the baseline average to count as confirmed. */
  volumeRatioThreshold: number;
  /** Window (in trading days, including the breakout day) used to check that volume was trending up into the breakout, not just a lone spike. */
  volumeTrendDays: number;
  /** How far back to look for the swing high (or, for a downward crossing, swing low) that makes `level` meaningful rather than arbitrary. */
  recentHighLookbackDays: number;
  /** `level` counts as "near the recent high" if within this fraction below the highest high (mirrored for a downward crossing: within this fraction above the lowest low). */
  recentHighTolerance: number;
  /** Number of subsequent trading days the close must stay past `level`, in the crossing direction, to call the move "held". */
  holdDays: number;
  /** Minimum bars of prior history required before we'll trust the baseline/recent-high calculations at all. */
  minBaselineBars: number;
}

export const DEFAULT_ANALYSIS_PARAMS: AnalysisParams = {
  baselineDays: 20,
  volumeRatioThreshold: 1.5,
  volumeTrendDays: 3,
  recentHighLookbackDays: 60,
  recentHighTolerance: 0.02,
  holdDays: 2,
  minBaselineBars: 10,
};

function findAlertBarIndex(bars: PriceBar[], alert: Alert): number | null {
  const alertDateStr = dateOnly(alert.time);
  for (let i = 0; i < bars.length; i++) {
    if (dateOnly(bars[i].date) >= alertDateStr) {
      return i;
    }
  }
  return null;
}

function dateOnly(d: Date): string {
  return d.toISOString().slice(0, 10);
}

function mean(values: number[]): number {
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function noDataVerdict(alert: Alert, reason: string): BreakoutVerdict {
  return {
    alert,
    closeOnAlertDay: null,
    pctAboveLevel: null,
    volumeOnAlertDay: null,
    avgVolumeBaseline: null,
    volumeRatio: null,
    volumeTrendRatio: null,
    nearRecentHigh: null,
    heldAboveLevel: null,
    daysHeld: 0,
    verdict: "INSUFFICIENT_DATA",
    notes: reason,
  };
}

export function analyzeAlert(
  alert: Alert,
  barsInput: PriceBar[],
  params: AnalysisParams = DEFAULT_ANALYSIS_PARAMS
): BreakoutVerdict {
  const bars = [...barsInput].sort((a, b) => a.date.getTime() - b.date.getTime());

  if (alert.alertType !== "price_cross" || alert.level === null) {
    return {
      alert,
      closeOnAlertDay: null,
      pctAboveLevel: null,
      volumeOnAlertDay: null,
      avgVolumeBaseline: null,
      volumeRatio: null,
      volumeTrendRatio: null,
      nearRecentHigh: null,
      heldAboveLevel: null,
      daysHeld: 0,
      verdict: "SKIPPED",
      notes: `Not a numeric price-level alert (${alert.alertType}).`,
    };
  }
  const level = alert.level;
  const down = alert.direction === "down";
  const past = down ? "below" : "above";
  const back = down ? "above" : "below";
  /** Whether a close is past the level in the crossing direction. */
  const closedPast = (close: number) => (down ? close < level : close > level);

  const idx = findAlertBarIndex(bars, alert);
  if (idx === null) {
    return noDataVerdict(alert, "No bar found on or after the alert date.");
  }

  const breakoutBar = bars[idx];
  const baseline = bars.slice(Math.max(0, idx - params.baselineDays), idx);
  if (baseline.length < params.minBaselineBars) {
    return noDataVerdict(
      alert,
      `Only ${baseline.length} prior trading day(s) of history available (need ${params.minBaselineBars}).`
    );
  }

  const avgVolumeBaseline = mean(baseline.map((b) => b.volume));
  const volumeRatio = avgVolumeBaseline > 0 ? breakoutBar.volume / avgVolumeBaseline : null;

  const trendDays = params.volumeTrendDays;
  const recentWindow = bars.slice(Math.max(0, idx - trendDays + 1), idx + 1);
  const priorWindow = bars.slice(Math.max(0, idx - 2 * trendDays + 1), Math.max(0, idx - trendDays + 1));
  let volumeTrendRatio: number | null = null;
  if (priorWindow.length > 0) {
    const priorAvg = mean(priorWindow.map((b) => b.volume));
    const recentAvg = mean(recentWindow.map((b) => b.volume));
    if (priorAvg > 0) {
      volumeTrendRatio = recentAvg / priorAvg;
    }
  }

  // The recent extreme on the side the crossing moved toward: the swing high
  // for an upward crossing, the swing low for a downward one.
  const extremeWindow = bars.slice(Math.max(0, idx - params.recentHighLookbackDays), idx);
  let nearRecentHigh = false;
  if (extremeWindow.length > 0) {
    if (down) {
      const recentLow = Math.min(...extremeWindow.map((b) => b.low));
      nearRecentHigh = level <= recentLow * (1 + params.recentHighTolerance);
    } else {
      const recentHigh = Math.max(...extremeWindow.map((b) => b.high));
      nearRecentHigh = level >= recentHigh * (1 - params.recentHighTolerance);
    }
  }

  const closeOnAlertDay = breakoutBar.close;
  const pctAboveLevel = ((closeOnAlertDay - level) / level) * 100;

  const following = bars.slice(idx + 1, idx + 1 + params.holdDays);
  let daysHeld = 0;
  for (const bar of following) {
    if (closedPast(bar.close)) {
      daysHeld += 1;
    } else {
      break;
    }
  }
  let heldAboveLevel: boolean | null;
  if (following.length < params.holdDays) {
    heldAboveLevel = null; // not enough time has passed yet to know
  } else {
    heldAboveLevel = daysHeld === params.holdDays;
  }

  const closedPastLevel = closedPast(closeOnAlertDay);
  const volumeConfirmed = volumeRatio !== null && volumeRatio >= params.volumeRatioThreshold;
  const volumeGrowing = volumeTrendRatio === null || volumeTrendRatio >= 1.0;

  const extreme = down ? "recent low" : "recent high";
  const notesParts: string[] = [
    `close ${closedPastLevel ? past : `at/${back}`} level (${pctAboveLevel >= 0 ? "+" : ""}${pctAboveLevel.toFixed(1)}%)`,
    volumeRatio !== null ? `volume ${volumeRatio.toFixed(2)}x ${params.baselineDays}d avg` : "volume ratio n/a",
  ];
  if (volumeTrendRatio !== null) {
    notesParts.push(`${trendDays}d volume trend ${volumeTrendRatio.toFixed(2)}x`);
  }
  notesParts.push(nearRecentHigh ? `near/${past} ${extreme}` : `well ${back} ${extreme} (arbitrary level)`);
  if (heldAboveLevel === true) {
    notesParts.push(`held ${past} level for ${daysHeld}/${params.holdDays}d`);
  } else if (heldAboveLevel === false) {
    notesParts.push(`failed to hold - closed back ${back} within ${params.holdDays}d`);
  } else {
    notesParts.push("not enough time elapsed yet to confirm it held");
  }

  let verdict: string;
  if (!closedPastLevel) {
    verdict = "NO_CLOSE_CONFIRM";
  } else if (volumeConfirmed && nearRecentHigh && volumeGrowing && heldAboveLevel !== false) {
    verdict = "CONFIRMED_BREAKOUT";
  } else if (volumeConfirmed && nearRecentHigh) {
    verdict = "WATCH";
  } else if (volumeConfirmed || nearRecentHigh) {
    verdict = "WATCH_WEAK";
  } else {
    verdict = "NO";
  }

  return {
    alert,
    closeOnAlertDay,
    pctAboveLevel,
    volumeOnAlertDay: breakoutBar.volume,
    avgVolumeBaseline,
    volumeRatio,
    volumeTrendRatio,
    nearRecentHigh,
    heldAboveLevel,
    daysHeld,
    verdict,
    notes: notesParts.join("; "),
  };
}
